python/ray/util/ml_utils/checkpoint_manager.py

```diff
@@ -0,0 +1,433 @@
+    (both in-memory and on-disk checkpoints). For on-disk checkpoints, it
+    keeps a configured number of checkpoints according to specified metrics.
+
+    The manager supports lazy data writing by utilizing the
+    ``TrackedCheckpoint.commit()`` API, which is only invoked if the checkpoint
+    should be persisted to disk.
+
+    Args:
+        checkpoint_strategy: Checkpoint strategy defining how many and which
+            checkpoints to keep.
+        latest_checkpoint_id: First checkpoint ID to use (e.g. in case we
+            continue training an existing experiment).
+        delete_fn: Function that takes a TrackedCheckpoint and deletes it from disk
+            or memory upon request.
+
+    """
+
+    # If memory checkpoints should be persisted
+    _persist_memory_checkpoints: bool = False
+
+    def __init__(
+        self,
+        checkpoint_strategy: CheckpointStrategy,
+        latest_checkpoint_id: int = 0,
+        delete_fn: Optional[Callable[["TrackedCheckpoint"], None]] = None,
+    ):
+        self._checkpoint_strategy = checkpoint_strategy or CheckpointStrategy()
+
+        # Incremental unique checkpoint ID of this run.
+        self._latest_checkpoint_id = latest_checkpoint_id
+
+        # Used for keeping top K checkpoints.
+        self._top_persisted_checkpoints: List[_HeapCheckpointWrapper] = []
+
+        # Best checkpoint altogether.
+        # Used for exposing best_checkpoint_path.
+        self._best_persisted_checkpoint: Optional[TrackedCheckpoint] = None
+        self._latest_persisted_checkpoint: Optional[TrackedCheckpoint] = None
+        self._latest_memory_checkpoint: Optional[TrackedCheckpoint] = None
+
+        # Checkpoints that are not immediately removed
+        self._checkpoints_to_clean_up = set()
+        self._delete_fn = delete_fn
+
+    def set_delete_fn(self, delete_fn: Optional[Callable[["TrackedCheckpoint"], None]]):
+        """Update the function called to delete persisted checkpoints.
+
+        Args:
+            delete_fn: Function that takes a tracked checkpoint as an argument and
+                deletes it from disk.
+        """
+        self._delete_fn = delete_fn
+
+    def register_checkpoint(self, checkpoint: TrackedCheckpoint):
+        """Register new checkpoint and add to bookkeeping.
+
+        This method will register a new checkpoint and add it to the internal
+        bookkeeping logic. This means the checkpoint manager will decide if
+        this checkpoint should be kept, and if older or worse performing
+        checkpoints should be deleted.
+
+        Args:
+            checkpoint: Tracked checkpoint object to add to bookkeeping.
+        """
+        checkpoint.id = checkpoint.id or self._latest_checkpoint_id
+
+        if checkpoint.storage_mode == CheckpointStorage.MEMORY:
+            self._replace_latest_memory_checkpoint(checkpoint)
+
+            if self._persist_memory_checkpoints:
+                persisted_checkpoint = copy.copy(checkpoint)
```

Comment:
do we need to copy here?

Reply:
Yes, as we're changing the storage mode. Otherwise the memory checkpoint (saved as last memory checkpoint) will have storage mode persistent.

Reply:
Note that the checkpoint will further alter its properties when calling `commit()`